README.md
poppac
======

*Personnal OID Printer Paper Check

>> What is it ?

Shinken module which takes as argument an OID and hostname.
It will bring into Shinken the total pages printed for this month and for the passed one.
It's regardless of brand or model due to the fact you can send the correct personnalized OID of your printer to this module.

>> Why to ?

To avoid the use of too bigger scripts which are not always stable with new Shinken/Naggios version.
Plus the fact that there were no really many scripts to get personnal information about paper type consummed directly inside the monitoring interface.
This script had been KISSed (Keep It Simply, Stupid).
It will no impact on future updates.

>> Could you be more complete about the features ?

* Shows the Total printed pages for the actual month plus the name of the month.
* Shows the total printed pages for the last month.
* Manage the month change automatically, based on system time.
* Detect and create files automatically at first launch.
* Detect and warn the user for connection problem to host.
* Detect and warn the user for corrupted or missing files.
* Detect system crash.

>> How does it work ?

Just call 
<<<<<<< HEAD
./poppac.pl OID HOSTNAME PAPERTYPE[A4B/A4C/A3B/A3C/...] (for French version)

OR

./poppac-en.pl OID HOSTNAME PAPERTYPE[A4B/A4C/A3B/A3C/...] (for English version)


CHANGELOG
#TODO
* Add graphical integration or with PNP4nagios
* Add install script for dependancies and folders creation
* Packaging for auto-install and configure

#Version 0.72
* Add error code into shinken UI [CRITICAL | WARNING]

#Version 0.71
* Add papertype selection
* Fix bad files creation with new file manager system 

#Version 0.7
* Add an History file of past month
* Script now exists in English - Poppac-en.pl
* Fix new year bug when passing from December to January

#Version 0.61
* Fix error about not right to create directory
* New error message send into Shinken

#Version 0.6
* Add own directory to store files
* Fix data restore in case of machine crash

#Version 0.5
* Add data save into files
* Add possibility to manually edit files
* Add automatic month change detection
* Add automatic data update on month change

#Version 0.41
* Add error message on deleted files

#Version 0.4
* Fix non-init data reception

#Version 0.3
* Data init and return at first launch
* Add detection and automake for missing files
* Add init detection

#Version 0.2
* Add the month for month check
* Add error management
* Fix multiple data send

#Version 0.1
* Simple OID information into the main output


poppac-en.pl
#!/usr/bin/perl -w

######## POPPAC - Personal OID Printer Paper Check ##############
# Version : 0.72
# Date :  March 27 2014
# Author  : Arnaud Comein (arnaud.comein@gmail.com)
# Licence : GPL - http://www.fsf.org/licenses/gpl.txt
#################################################################

#Besoin pour les GET SNMP
use BER;
use SNMP_util;
use SNMP_Session;

#Besoin pour le debug - ~mode verbeux sous linux
use strict;
use warnings;

#Variables - "shift" = attente d'argument - !ordre! - Undef par defaut
my $MIB = shift;
my $HOST = shift;
my $FILE = shift;
my $dir = "/usr/share/poppac";
my $O;
my $value;
my $tot;
my $lastTot;
my $i = 0;
my $totAll;
my $monthRaed;
my $INI;
my $MONTH;
my $HIST;
my $TOT;
my $LAST;
my %ERRORS=('OK'=>0,'WARNING'=>1,'CRITICAL'=>2,'UNKNOWN'=>3,'DEPENDENT'=>4);

#Déclaration pour récupération de la date
my ($second, $minute, $hour, $dayOfMonth, $month, $yearOffset, $dayOfWeek, $dayOfYear, $daylightSavings) = localtime();

#Ajustement de l'année (la date commence à 1900)
my $year = $yearOffset + 1900;

#Récupération du mois en cours - Janvier = 0 en systeme
my $monthnum = $month + 1; #Besoin pour l'historique

my %monthname = (
1 => 'Janvier',	
2 => 'Fevrier',
3 => 'Mars',
4 => 'Avril',
5 => 'Mai',
6 => 'Juin',
7 => 'Juillet',
8 => 'Aout',
9 => 'Septembre',
10 => 'Octobre',
11 => 'Novembre',
12 => 'Decembre',
);

#Centralisation des erreurs
my $err = "Please, remove all *-init.txt files inside /usr/share/poppac";
my $help = "Correct use : ./poppac.pl OID HOSTNAME PAPERTYPE[A4B/A4C/A3B/A3C/...]\n";
my $errfile = "Plaese, create directory /usr/share/poppac and give rights on it to your user";
my $errcon = "Unable to connect to $HOST, Check your OID number and your hostname\n";

#Help
($MIB) && ($HOST) && ($FILE) || die $help;

#GetOID
($value) = &snmpget("public\@$HOST","$MIB");

#Prevoir une sortie si la connexion à l'hote ne se fait pas
if ($value)
{
	#Est-ce la premiere éxécution ?
	open $INI, "<", "/usr/share/poppac/$HOST-$FILE-init.txt" or $i = 1;

	if ($i == 1) #Si le fichier init n'existe pas, on crée les fichiers et les ouvres en écriture.
	{	
		mkdir $dir;
		
		open $INI, ">", "/usr/share/poppac/$HOST-$FILE-init.txt" or print $errfile and die;
		open $MONTH, ">", "/usr/share/poppac/$HOST-$FILE-month.txt";
		open $TOT, ">", "/usr/share/poppac/$HOST-$FILE-tot.txt";
		open $LAST, ">", "/usr/share/poppac/$HOST-$FILE-last.txt";		
		open $HIST, ">", "/usr/share/poppac/$HOST-$FILE-hist.txt";

		$totAll = 0;
		$lastTot = $value;
		$monthRaed = $monthnum;
	}
	else #sinon, on les ouvre en lecture.
	{	
		open $MONTH, "<", "/usr/share/poppac/$HOST-$FILE-month.txt" or print $err and die;
		open $TOT, "<", "/usr/share/poppac/$HOST-$FILE-tot.txt" or print $err and die;
		open $LAST, "<", "/usr/share/poppac/$HOST-$FILE-last.txt" or print $err and die;	

		$monthRaed = <$MONTH>;
		$totAll = <$TOT>;
		$lastTot = <$LAST>;
	}

	#Calcul du total du mois en cours = Valeur SNMP - dernier mois - Totalhistorique
	$tot = ($value - $lastTot - $totAll);

	#Verification si changement de mois
	if ($monthRaed != $monthnum)
	{
		$totAll = $totAll + $lastTot;
		$lastTot = $tot;
		$tot = ($value - $lastTot - $totAll);
		$i = 2;
	}

	#Retour Shinken WebUI
	print "Printed pages for $monthname{$monthnum} : $tot * Last month : $lastTot\n";

	#Ecriture des nouvelles données dans les fichiers
	if ($i == 1) 
	{ 
		print $INI $i; #Ecriture ssi init viens d'etre créé
		print $MONTH $monthnum;
		print $TOT $totAll;
		print $LAST $lastTot;
	}	
	#Ecriture ssi le mois viens de changer. +> pour ajouter ecriture	
	if ($i == 2)
	{ 	
		open $MONTH, "+>", "/usr/share/poppac/$HOST-$FILE-month.txt";
		open $TOT, "+>", "/usr/share/poppac/$HOST-$FILE-tot.txt";
		open $LAST, "+>", "/usr/share/poppac/$HOST-$FILE-last.txt";	
		open $HIST, "+>>", "/usr/share/poppac/$HOST-$FILE-hist.txt";

		print $MONTH $monthnum; 
		print $TOT $totAll;
		print $LAST $lastTot;
		
		#Gestion du passage de Décembre à Janvier sinon normal
		if ($month == 0)
		{ print $HIST "$monthname{12} : $lastTot\n***** New Year : $year *****\n"; }
		else { print $HIST "$monthname{$month} : $lastTot\n"; }
	}

	#Fermeture des fichiers
	close $INI;
	close $MONTH;
	close $TOT;
	close $LAST;
	if ($i == 2) { close $HIST; }

} #Fin de la sortie en cas d'erreur de connexion

else 
{ 
	print $errcon;
	exit $ERRORS{"CRITICAL"};
}


poppac.pl
#!/usr/bin/perl -w

######## POPPAC - Personal OID Printer Paper Check ##############
# Version : 0.72
# Date :  March 27 2014
# Author  : Arnaud Comein (arnaud.comein@gmail.com)
# Licence : GPL - http://www.fsf.org/licenses/gpl.txt
#################################################################

#Besoin pour les GET SNMP
use BER;
use SNMP_util;
use SNMP_Session;

#Besoin pour le debug - ~mode verbeux sous linux
use strict;
use warnings;

#Variables - "shift" = attente d'argument - !ordre! - Undef par defaut
my $MIB = shift;
my $HOST = shift;
my $FILE = shift;
my $dir = "/usr/share/poppac";
my $O;
my $value;
my $tot;
my $lastTot;
my $i = 0;
my $totAll;
my $monthRaed;
my $INI;
my $MONTH;
my $HIST;
my $TOT;
my $LAST;
my %ERRORS=('OK'=>0,'WARNING'=>1,'CRITICAL'=>2,'UNKNOWN'=>3,'DEPENDENT'=>4);

#Déclaration pour récupération de la date
my ($second, $minute, $hour, $dayOfMonth, $month, $yearOffset, $dayOfWeek, $dayOfYear, $daylightSavings) = localtime();

#Ajustement de l'année (la date commence à 1900)
my $year = $yearOffset + 1900;

#Récupération du mois en cours - Janvier = 0 en systeme
my $monthnum = $month + 1; #Besoin pour l'historique

my %monthname = (
1 => 'Janvier',	
2 => 'Fevrier',
3 => 'Mars',
4 => 'Avril',
5 => 'Mai',
6 => 'Juin',
7 => 'Juillet',
8 => 'Aout',
9 => 'Septembre',
10 => 'Octobre',
11 => 'Novembre',
12 => 'Decembre',
);

#Centralisation des erreurs
my $err = "Veuillez supprimer les fichiers *-init.txt se trouvant dans /usr/share/poppac\n";
my $help = "Utilisation : ./poppac.pl OID HOSTNAME PAPERTYPE[A4B/A4C/A3B/A3C/...]\n";
my $errfile = "Veuillez créer le dossier /usr/share/poppac\n";
my $errcon = "Impossible d'établir la connexion avec $HOST, Verifiez le numéro OID et le nom d'hôte\n";

#Help
($MIB) && ($HOST) && ($FILE) || die $help;

#GetOID
($value) = &snmpget("public\@$HOST","$MIB");

#Prevoir une sortie si la connexion à l'hote ne se fait pas
if ($value)
{
	#Est-ce la premiere éxécution ?
	open $INI, "<", "/usr/share/poppac/$HOST-$FILE-init.txt" or $i = 1;

	if ($i == 1) #Si le fichier init n'existe pas, on crée les fichiers et les ouvres en écriture.
	{	
		mkdir $dir;
		
		open $INI, ">", "/usr/share/poppac/$HOST-$FILE-init.txt" or print $errfile and die;
		open $MONTH, ">", "/usr/share/poppac/$HOST-$FILE-month.txt";
		open $TOT, ">", "/usr/share/poppac/$HOST-$FILE-tot.txt";
		open $LAST, ">", "/usr/share/poppac/$HOST-$FILE-last.txt";		
		open $HIST, ">", "/usr/share/poppac/$HOST-$FILE-hist.txt";

		$totAll = 0;
		$lastTot = $value;
		$monthRaed = $monthnum;
	}
	else #sinon, on les ouvre en lecture.
	{	
		open $MONTH, "<", "/usr/share/poppac/$HOST-$FILE-month.txt" or print $err and die;
		open $TOT, "<", "/usr/share/poppac/$HOST-$FILE-tot.txt" or print $err and die;
		open $LAST, "<", "/usr/share/poppac/$HOST-$FILE-last.txt" or print $err and die;	

		$monthRaed = <$MONTH>;
		$totAll = <$TOT>;
		$lastTot = <$LAST>;
	}

	#Calcul du total du mois en cours = Valeur SNMP - dernier mois - Totalhistorique
	$tot = ($value - $lastTot - $totAll);

	#Verification si changement de mois
	if ($monthRaed != $monthnum)
	{
		$totAll = $totAll + $lastTot;
		$lastTot = $tot;
		$tot = ($value - $lastTot - $totAll);
		$i = 2;
	}

	#Retour Shinken WebUI
	print "Impressions pour $monthname{$monthnum} : $tot * Mois dernier : $lastTot\n";

	#Ecriture des nouvelles données dans les fichiers
	if ($i == 1) 
	{ 
		print $INI $i; #Ecriture ssi init viens d'etre créé
		print $MONTH $monthnum;
		print $TOT $totAll;
		print $LAST $lastTot;
	}	
	#Ecriture ssi le mois viens de changer. +> pour ajouter ecriture	
	if ($i == 2)
	{ 	
		open $MONTH, "+>", "/usr/share/poppac/$HOST-$FILE-month.txt";
		open $TOT, "+>", "/usr/share/poppac/$HOST-$FILE-tot.txt";
		open $LAST, "+>", "/usr/share/poppac/$HOST-$FILE-last.txt";	
		open $HIST, "+>>", "/usr/share/poppac/$HOST-$FILE-hist.txt";

		print $MONTH $monthnum; 
		print $TOT $totAll;
		print $LAST $lastTot;
		
		#Gestion du passage de Décembre à Janvier sinon normal
		if ($month == 0)
		{ print $HIST "$monthname{12} : $lastTot\n***** Nouvelle année : $year *****\n"; }
		else { print $HIST "$monthname{$month} : $lastTot\n"; }
	}

	#Fermeture des fichiers
	close $INI;
	close $MONTH;
	close $TOT;
	close $LAST;
	if ($i == 2) { close $HIST; }

} #Fin de la sortie en cas d'erreur de connexion

else 
{ 
	print $errcon;
	exit $ERRORS{"CRITICAL"};
}
